src/package/packagePushUpgrade.ts

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import util from 'node:util';
+import { Connection, SfError, SfProject } from '@salesforce/core';
+import { Schema, QueryResult } from '@jsforce/jsforce-node';
+import { IngestJobV2, IngestJobV2FailedResults } from '@jsforce/jsforce-node/lib/api/bulk2';
+import {
+  PackagePushRequestListQueryOptions,
+  PackagePushRequestListResult,
+  PackagePushScheduleResult,
+  PackagePushRequestReportQueryOptions,
+  PackagePushRequestReportResult,
+  PackagePushRequestJobCountByStatusResult,
+  PackagePushRequestReportJobFailuresResult,
+  PackagePushRequestAbortQueryOptions,
+} from '../interfaces';
+import { applyErrorAction, massageErrorMessage } from '../utils/packageUtils';
+
+export type PackagePushRequestListOptions = {
+  connection: Connection;
+  packageId: string;
+  project?: SfProject;
+};
+
+type PackagePushRequestResult = {
+  id: string;
+  success: boolean;
+  errors: object[];
+};
+
+export class PackagePushUpgrade {
+  public constructor() {}
+
+  public static async list(
+    connection: Connection,
```

Comment:
NIT: put connection as a class property to avoid having to pass it in everywhere